### ISSUE TITLE
[XPU] SYCL assert in TensorCompareKernels aborts DeltaNet reference fallback (div-by-zero instead of inf/NaN)

### ISSUE THREAD

**Issue by baustin27** [pytorch/pytorch#197777]
# Bug Report: Intel XPU DeltaNet reference kernel assertion failure in torch_chunk_gated_delta_rule

## Environment
- **Hardware**: Intel Arc Pro B70 Graphics (32GB VRAM)
- **OS**: Ubuntu 24.04 (Proxmox CT)
- **Compute Runtime**: 26.35.39758.10
- **PyTorch**: 2.14.0+xpu
- **Transformers**: 5.17.0 (includes `qwen3_5_moe` modeling)
- **Flash-linear-attention**: NOT installed (falls back to reference PyTorch implementation)

## Description
When running Qwen3.6-35B-A3B (a Gated DeltaNet hybrid architecture) on XPU, the model load succeeds but the first forward pass triggers an assertion failure in the reference implementation of `chunk_gated_delta_rule` (PyTorch fallback for `flash-linear-attention` which has no XPU kernels).

The assertion is in the SYCL tensor comparison kernel:
```
/__w/pytorch/pytorch/third_party/torch-xpu-ops/src/ATen/native/xpu/sycl/TensorCompareKernels.cpp:193: operator(): global id: [0,0,0], local id: [0,0,0] Assertion `input_[0] != 0` failed.
```

This indicates a divide-by-zero in the reference kernel's triangular solve or l2norm path.

## Minimal Repro
```python
import torch
from transformers import AutoModelForCausalLM
from gptqmodel import GPTQModel, BACKEND, QuantizeConfig

# Use Qwen3.6 GPTQ-Int4 weights (any quantized Qwen3.6 works)
G = '/models/hub/models--palmfuture--Qwen3.6-35B-A3B-GPTQ-Int4/snapshots/00a66983516f8f8057741221277eed4141c9e431'
qc = QuantizeConfig.from_pretrained(G)
qc.device = 'xpu:0'
m = GPTQModel.from_pretrained(G, quantize_config=qc, backend=BACKEND.GPTQ_TORCH)

tok = m.tokenizer
p = tok.apply_chat_template([{'role':'user','content':'Reply with exactly one word: Paris'}], tokenize=False, add_generation_prompt=True)
ids = tok(p, return_tensors='pt').to('xpu:0')

with torch.no_grad():
    m.model(**ids)  # CRASHES HERE with SYCL assertion
```

## Stack Trace (abridged)
```
File ".../transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py", line 641
    core_attn_out, last_recurrent_state = torch_chunk_gated_delta_rule(
File ".../transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py", line 302
    @use_kernel_func_from_hub_with_fallback("chunk_gated_delta_rule", "fla")
    def torch_chunk_gated_delta_rule(...):
File ".../torch/native/xpu/sycl/TensorCompareKernels.cpp", line 193
    Assertion `input_[0] != 0` failed.
```

## Root Cause Analysis
The reference `torch_chunk_gated_delta_rule` in transformers performs:
1. `l2norm(query, dim=-1)` → `rsqrt(sum(x^2) + eps)`
2. Triangular solve: `torch.linalg.solve_triangular(ut_system, v_beta, upper=False, unitriangular=True)`
3. Cumulative decay: `cum_decay = decay.cumsum(dim=3)` then `cum_decay.exp()`

Any of these can produce zero denominators. The XPU SYCL implementation of `rsqrt`/`solve_triangular`/`exp` appears to hit an assert when input is zero, whereas CUDA silently produces inf/NaN.

## Workarounds Tried
1. **Install flash-linear-attention** → fails because it imports `torch.cuda` internally (CUDA-only)
2. **Force fp32 compute** → PyTorch XPU fp32 path works, but model is fp16 quantized; runtime conversion too slow
3. **Disable DeltaNet layers** → not possible without model surgery

## Impact
- Blocks all Gated DeltaNet hybrid models (Qwen3.6, Nemotron 3.5 Lightning, future Qwen3.8) on XPU
- The reference implementation is the *only* path since flash-linear-attention has no XPU backend
- Makes B70 unusable for the entire DeltaNet architecture class

## Request
1. **Fix the SYCL assertion** in `TensorCompareKernels.cpp` - zero input should produce inf/NaN per IEEE 754, not crash
2. **Provide XPU kernels for flash-linear-attention** (chunk_gated_delta_rule, fused_recurrent_gated_delta_rule)
3. **Document the fp32 requirement** for DeltaNet on XPU until kernels exist

## Additional Context
- The same model runs correctly on CPU (fp32 reference) and CUDA (flash-linear-attention optimized kernels)
- The bug is specific to the XPU reference fallback path
- Affected architectures: Qwen3.5 MoE, Qwen3.6, Nemotron 3.5 Lightning (Mamba-2 hybrid)

cc @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by aron-intframe** [comment]
The assertion at `TensorCompareKernels.cpp:193` is not a divide-by-zero check inside rsqrt/exp/solve_triangular. That line is the body of `AssertAsyncKernelFunctor1`, i.e. the kernel behind `torch._assert_async`: it does nothing except verify that a tensor someone passed to `_assert_async` is non-zero. The message argument is currently dropped on XPU (the `TODO` right above it points to intel/torch-xpu-ops#5165), which is why you only see `input_[0] != 0` and not what was being asserted.

Two things follow:

1. The ops you listed cannot produce this. In fp32 they yield inf/nan/0, and the triangular solve in `torch_chunk_gated_delta_rule` is `unitriangular=True`, so the diagonal is never even read. Checked on 2.14.0+cpu (same numerics; none of these ops have a device assert on any backend):

```
$ python3 repro.py
torch 2.14.0+cpu
l2norm rsqrt(sum(x^2)+eps) on zero rows -> [999.9999389648438, 999.9999389648438]
rsqrt(0) -> inf
exp(-inf) -> 0.0
solve_triangular(unitriangular=True) with all-zero A -> finite: True (diagonal is never read)
solve_triangular(unitriangular=False) with zero diag -> nan count: 504 , no exception
torch._assert_async(0) on CPU -> Expected Tensor with single nonzero value, but got zero
index_put_(accumulate=True) OOB on CPU -> index 7 is out of bounds for dimension 0 with size 4
```

2. The Python frame in your trace is the first host sync after the bad kernel was queued, not the op that failed. `solve_triangular` on XPU without oneMKL is a CPU fallback (`BatchLinearAlgebra.cpp`, `triangular_solve_kernel_fallback` copies A and B to CPU), so it is the first blocking point in the forward pass; if your log shows the warning `torch.linalg.solve_triangular op is using fallback implementation`, that is exactly what happened. The offending op ran earlier in the same layer or before it (embedding, norms, the GPTQ dequant, conv1d).

What calls `_assert_async` in eager mode and can land on that line:

- `index_put_(..., accumulate=True)` (or any `index_put_` under `torch.use_deterministic_algorithms(True)`) with an out-of-range index. On XPU the range check is `at::_assert_async(index_max < dim_size)` / `at::_assert_async(index_min >= -dim_size)` in `wrapIndexOnce` (torch-xpu-ops `sycl/IndexingUtils.h`); CUDA does the same in `Indexing.cu`. CPU checks synchronously and names the index, as in the last output line above.
- `torch.multinomial` (invalid probabilities) and `torch._assert_async` called directly. Note that `torch.compile` rewrites Python `assert` on tensors into `_assert_async`, and GPTQModel's `TorchQuantLinear.optimize()` compiles `dequantize_weight` by default, so a failed assert inside the dequant path is a candidate too.

Minimal repro without the model or the checkpoint, which is what the triage comment asked for. Both should print the identical `Assertion input_[0] != 0 failed` at line 193 (I do not have an XPU, so I could not run these myself):

```python
import torch
torch._assert_async(torch.zeros((), device="xpu"))
torch.xpu.synchronize()
```

```python
import torch
x = torch.zeros(4, device="xpu")
x.index_put_((torch.tensor([7], device="xpu"),), torch.ones(1, device="xpu"), accumulate=True)
torch.xpu.synchronize()
```

To find which op in your model queues it, synchronize after every op so the traceback lands on the real caller:

```python
from torch.overrides import TorchFunctionMode
class SyncEveryOp(TorchFunctionMode):
    def __torch_function__(self, func, types, args=(), kwargs=None):
        out = func(*args, **(kwargs or {}))
        torch.xpu.synchronize()
        return out

with torch.no_grad(), SyncEveryOp():
    m.model(**ids)
```

Run it once with `optimize`/torch.compile disabled on the GPTQ layers as well, so a `_assert_async` from a compiled dequant is not hidden behind a graph. If the failing op turns out to be an `index_put_`, print `index.min()/max()` and the dim size right before it; the value that shows up (an expert id, a `g_idx` entry, a token id) tells you whether this is a checkpoint/packing problem or an XPU kernel bug. Request 1 in the issue ("zero input should produce inf/NaN") does not apply to this kernel; the actionable XPU-side item is #5165 (propagate the assert message) so this class of failure stops being opaque.

**Comment by EikanWang** [comment]
@chuanqi129 , @ZhaoqiongZ , pls. prioritize the issue.

**Comment by baustin27** [comment]
Sorry guys, I think my AI made all that. I definitely didn't put this in
here. If it's wrong at all, my assistant must have submitted that.

Bobby Austin
Mobile:972-322-8575
CC# 712-775-7035
Entry Code:949078

**Comment by CuiYifeng** [comment]
@baustin27 Thanks for the clarification.
The assertion in the “Minimal repro without the model or the checkpoint” is expected behavior. Can we close this issue?
